graylog2-server/src/main/java/org/graylog2/indexer/results/ScrollResult.java

```diff
@@ -16,73 +16,106 @@
-import org.elasticsearch.search.SearchHits;
+import org.graylog2.indexer.searches.SearchException;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.io.IOException;
 import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
 
 public class ScrollResult extends IndexQueryResult {
     private static final Logger LOG = LoggerFactory.getLogger(ScrollResult.class);
 
-    private final Client client;
+    public interface Factory {
+        ScrollResult create(SearchResult initialResult, String query, List<String> fields);
+    }
+
+    private final JestClient jestClient;
+    private final ObjectMapper objectMapper;
+    private SearchResult initialResult;
     private final List<String> fields;
     private final String queryHash; // used in log output only
     private final long totalHits;
-    private SearchResponse firstResponse;
 
     private String scrollId;
     private int chunkId = 0;
 
-    public ScrollResult(Client client,
-                        String originalQuery,
-                        BytesReference builtQuery,
-                        SearchResponse response, List<String> fields) {
-        super(originalQuery, builtQuery, response.getTook());
-        this.client = client;
+    @AssistedInject
+    public ScrollResult(JestClient jestClient, ObjectMapper objectMapper, @Assisted SearchResult initialResult, @Assisted String query, @Assisted List<String> fields) {
+        super(query, null, new TimeValue(initialResult.getJsonObject().get("took").getAsLong()));
+        this.jestClient = jestClient;
+        this.objectMapper = objectMapper;
+        this.initialResult = initialResult;
         this.fields = fields;
-        totalHits = response.getHits().totalHits();
-        scrollId = response.getScrollId();
-        firstResponse = response;
+        totalHits = initialResult.getTotal();
+        scrollId = getScrollIdFromResult(initialResult);
 
         final Md5Hash md5Hash = new Md5Hash(getOriginalQuery());
         queryHash = md5Hash.toHex();
 
         LOG.debug("[{}] Starting scroll request for query {}", queryHash, getOriginalQuery());
     }
 
-    public ScrollChunk nextChunk() {
+    public ScrollChunk nextChunk() throws IOException {
 
-        final SearchResponse search;
+        final JestResult search;
         // make sure to return the initial hits, see https://github.com/Graylog2/graylog2-server/issues/2126
-        if (firstResponse == null) {
-            search = client.prepareSearchScroll(scrollId)
-                    .setScroll(TimeValue.timeValueMinutes(1))
-                    .execute()
-                    .actionGet();
+        final List<ResultMessage> hits;
+        if (initialResult == null) {
+            search = getNextScrollResult();
+            hits = StreamSupport.stream(search.getJsonObject().getAsJsonObject("hits").getAsJsonArray("hits").spliterator(), false)
```

Comment:
This will fail in various ways (`NullPointerException`, `IllegalArgumentException`, `ClassCastException`) if the response JSON isn't anything like it's expected here.
Maybe take a look at the `GsonUtils` class.